controllers/etcdcluster_controller.go

```diff
@@ -62,6 +68,83 @@ func headlessServiceForCluster(cluster *etcdv1alpha1.EtcdCluster) *v1.Service {
 	}
 }
 
+// updateStatus updates the EtcdCluster resource's status to be the current value of the cluster.
+func (r *EtcdClusterReconciler) updateStatus(
+	ctx context.Context,
+	cluster *etcdv1alpha1.EtcdCluster,
+	members *[]etcdclient.Member,
+	reconcilerEvent reconcilerevent.ReconcilerEvent) error {
+
+	if members != nil {
+		cluster.Status.Members = make([]etcdv1alpha1.EtcdMember, len(*members))
+		for i, member := range *members {
+			cluster.Status.Members[i] = etcdv1alpha1.EtcdMember{
+				Name: member.Name,
+				ID:   member.ID,
+			}
+		}
+
+		if err := r.Client.Status().Update(ctx, cluster); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *EtcdClusterReconciler) reconcile(
+	ctx context.Context,
+	members *[]etcdclient.Member,
+	cluster *etcdv1alpha1.EtcdCluster,
+) (
+	ctrl.Result,
+	reconcilerevent.ReconcilerEvent,
+	error) {
+
+	name := types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}
+	log := r.Log.WithValues("cluster", name)
+
+	service := &v1.Service{}
+	if err := r.Get(ctx, name, service); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil, fmt.Errorf("unable to fetch EtcdCluster service: %w", err)
+		}
+		service = headlessServiceForCluster(cluster)
+		if err := r.Create(ctx, service); err != nil {
+			return ctrl.Result{}, nil, fmt.Errorf("unable to create service: %w", err)
+		}
+		log.V(1).Info("Created Service", "service", service.Name)
+		return ctrl.Result{}, &reconcilerevent.ServiceCreatedEvent{Object: cluster, ServiceName: service.Name}, nil
+	}
+	log.V(2).Info("Service exists", "service", service.Name)
+
+	if members == nil {
+		peers := &etcdv1alpha1.EtcdPeerList{}
+		if err := r.List(ctx, peers, client.MatchingFields{clusterNameSpecField: cluster.Name}); err != nil {
+			return ctrl.Result{}, nil, fmt.Errorf("unable to list peers: %w", err)
+		}
+
+		if cluster.Spec.Replicas != nil && int32(len(peers.Items)) < *cluster.Spec.Replicas {
+			// Create more peers
+			peerName := nextAvailablePeerName(cluster, peers.Items)
+			log.V(1).Info("Insufficient peers for replicas, adding new peer",
+				"current-peers", len(peers.Items),
+				"desired-peers", cluster.Spec.Replicas,
+				"peer", peerName)
+			peer := peerForCluster(cluster, peerName)
+			if err := r.Create(ctx, peer); err != nil {
+				return ctrl.Result{}, nil, fmt.Errorf("failed to create EtcdPeer %s: %w", peerName, err)
+			}
+			return ctrl.Result{}, &reconcilerevent.PeerCreatedEvent{Object: cluster, PeerName: peer.Name}, nil
+		}
+	} else {
+		log.V(2).Info("Cluster communication established")
+	}
+	return ctrl.Result{RequeueAfter: time.Second * 10}, nil, nil
```

Comment:
Why re-queue?  Is this to health check the etcdcluster every 10 seconds?

Reply:
Yes that's correct. We don't have a watch yet so we need a way of knowing of the member list has changed.